samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java

```diff
@@ -186,13 +190,20 @@ public void append(LogEvent event) {
       if (!systemInitialized) {
         // configs are needed to set up producer system, so check that before actually initializing
         if (this.loggingContextHolder.getConfig() != null) {
-          synchronized (this) {
-            if (!systemInitialized) {
-              setupSystem();
-              systemInitialized = true;
+          if (setUpSystemLock.tryLock(SET_UP_SYSTEM_TIMEOUT_MILLI_SECONDS, TimeUnit.MILLISECONDS)) {
+            try {
+              if (!systemInitialized) {
+                setupSystem();
```

Comment:
Is this supposed to be run only once? I am thinking if we should also set it to `true` within finally block? 
Alternatively, use `AtomicBoolean` and check and set in ln:190.

Reply:
Yes, it is supposed to be run only once.
I prefer not to set it to true within finally block because we want to make sure it is only set to true if setupSystem() is completed successfully. In case that setupSystem() fails (throwing some exceptions for example), we want to leave it as false.

The same reason applies to the option of AtomicBoolean. there are two cons for that:
1. same as above, can not make sure setupSystem() runs fine
2. perf reason, compareAndSet AtomicBoolean in line 190 means every time LOG.xxx() StreamAppender needs to check the AtomicBoolean which is expensive than checking a boolean

Reply:
There are no guarantees from `setupSystem` on providing idempotence nor able to setup system on the subsequent retries in case of failures.

So if we are handling externally for the former, why not handle the latter as well.

Reply:
I am not sure where we landed here. Can you update the comment with our discussion?

Reply:
Based on the discussion offline:
Yes, `setupSystem` does not provide idempotence nor retry logic to setup system upon failures.

The current implementation handles both by putting 
setupSystem() and  systemInitialized = true 
into a critical section to ensure that
1. no concurrent calls to setupSystem()
2. If setupSystem() completes, it won't be called again
3. If setupSystem() fails, it will be invoked by another thread that enters the critical section

However, 2 and 3 might not be the ideal way to handle idempotence and recovery from failures. We should revisit their reliabilities and consider other solutions 

one preferred option is that:
setupSystem() should handle idempotence and recovery from failures by itself and StreamAppender doesn't need to handle them explicitly in append().